README.md
# Zwerm Bot Configuration Schema

JSON schema of verbose configuration settings for bot instances.

`bot.schema.json` defines the schema for the configuration details for a single
bot instance.  
`bots.schema.json` defines the schema for a map of bot ids to their respective
configuration details for multiple bot instances.

### Installing

Install all dependencies:

```
npm install

```

## Release

To manually publish a version, run

```
npm publish --access public
```

### Versioning

We follow [SemVer](http://semver.org/) for versioning.

When making a Pull Request, add your changes to the _Unreleased_ section of
[CHANGELOG](https://github.com/zwerm/bot-configs-schema/blob/master/CHANGELOG.md).

To release a new version, run the `npm version` command on `master`, it will
take care of everything for you.

`npm version [major | minor | patch]`

## Development

### Branching model

We use a modified
[GitFlow](https://datasift.github.io/gitflow/IntroducingGitFlow.html) branching
model. The main difference being is we use `master` as a combined development
branch and branch of short living production branches as needed.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/)
and this project adheres to
[Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [3.3.0] - 2023-06-12

- Generates types for bot schemas

## [3.2.1] - 2023-05-29

- Updates for OpenAI engine schema

## [3.2.0] - 2023-05-15

## [3.1.0] - 2020-04-22

### Changed

- Update regions for aws services.
- Update voices for aws polly.

## [3.0.0] - 2018-10-24

### Added

- Support for IBM Watson.

## [2.3.0] - 2018-10-19

### Added

- Optional `'icon'` property to `webchat-channel-settings` object property.

## [2.2.0] - 2018-09-03

### Added

- Optional `'icon'` property to `options` object property.
- Optional `'supportedSystemEvents'` property to `options` object property.
- Slack configuration support

## [2.1.0] - 2018-07-17

### Added

- `us-west-2` is now a valid Amazon Lex regions.

## [2.0.1] - 2018-06-20

Bug fix release for the php class trying to include the schema from the wrong
path.

### Changed

- Use php constants to for schema file path.

## [2.0.1] - 2018-06-20

Maintenance version to fix php path.

## [2.0.0] - 2018-06-20

This major release has the `stamp` channel renamed to `api`. This is just a name
change; no other rules have been changed.

### Changed

- Renamed `stamp` channel to `api`. ([277d75a])
- Regenerated schema typings. ([f135e9a])
- Updated PHP class to use namespacing ([3e40f80])

## [1.0.0] - 2018-06-18

Official Release

## 0.1.0 - 2018-06-18

### Added

- Initial commit

<!-- prettier-ignore-start -->

[Unreleased]: https://github.com/zwerm/bot-configs-schema/compare/v2.0.1...master

[2.0.1]: https://github.com/zwerm/bot-configs-schema/compare/v2.0.0...v2.0.1
[2.0.0]: https://github.com/zwerm/bot-configs-schema/compare/v1.0.0...v2.0.0
[1.0.0]: https://github.com/zwerm/bot-configs-schema/compare/v0.1.0...v1.0.0

[277d75a]: https://github.com/zwerm/bot-configs-schema/commit/277d75a
[f135e9a]: https://github.com/zwerm/bot-configs-schema/commit/f135e9a
[3e40f80]: https://github.com/zwerm/bot-configs-schema/commit/3e40f80

<!-- prettier-ignore-end -->


bot.schema.d.ts
export declare namespace BotsSchema {
  /* tslint:disable */
  /**
   * This file was automatically generated by json-schema-to-typescript.
   * DO NOT MODIFY IT BY HAND. Instead, modify the source JSONSchema file,
   * and run json-schema-to-typescript to regenerate this file.
   */

  /**
   * This interface was referenced by `BotConfig`'s JSON-Schema
   * via the `definition` "bot-engine".
   */
  export type BotEngine =
    | AwsLexEngine
    | DialogflowEngine
    | WatsonEngine
    | OpenaiEngine;
  /**
   * This interface was referenced by `BotConfig`'s JSON-Schema
   * via the `definition` "adapter-label".
   */
  export type AdapterLabel = string | null;
  /**
   * This interface was referenced by `BotConfig`'s JSON-Schema
   * via the `definition` "aws-region-lex".
   */
  export type AwsRegionLex =
    | 'us-east-1'
    | 'us-west-2'
    | 'eu-west-1'
    | 'ap-southeast-2';
  /**
   * This interface was referenced by `BotConfig`'s JSON-Schema
   * via the `definition` "dialogflow-engine-settings".
   */
  export type DialogflowEngineSettings =
    | DialogflowEngineSettingsV1
    | DialogflowEngineSettingsV2;
  /**
   * This interface was referenced by `BotConfig`'s JSON-Schema
   * via the `definition` "watson-engine-settings".
   */
  export type WatsonEngineSettings =
    | WatsonEngineSettingsV1
    | WatsonEngineSettingsV2;
  /**
   * This interface was referenced by `BotConfig`'s JSON-Schema
   * via the `definition` "aws-polly-voice".
   */
  export type AwsPollyVoice =
    | 'Aditi'
    | 'Amy'
    | 'Astrid'
    | 'Bianca'
    | 'Brian'
    | 'Camila'
    | 'Carla'
    | 'Carmen'
    | 'Celine'
    | 'Chantal'
    | 'Conchita'
    | 'Cristiano'
    | 'Dora'
    | 'Emma'
    | 'Enrique'
    | 'Ewa'
    | 'Filiz'
    | 'Geraint'
    | 'Giorgio'
    | 'Gwyneth'
    | 'Hans'
    | 'Ines'
    | 'Ivy'
    | 'Jacek'
    | 'Jan'
    | 'Joanna'
    | 'Joey'
    | 'Justin'
    | 'Karl'
    | 'Kendra'
    | 'Kimberly'
    | 'Lea'
    | 'Liv'
    | 'Lotte'
    | 'Lucia'
    | 'Lupe'
    | 'Mads'
    | 'Maja'
    | 'Marlene'
    | 'Mathieu'
    | 'Matthew'
    | 'Maxim'
    | 'Mia'
    | 'Miguel'
    | 'Mizuki'
    | 'Naja'
    | 'Nicole'
    | 'Penelope'
    | 'Raveena'
    | 'Ricardo'
    | 'Ruben'
    | 'Russell'
    | 'Salli'
    | 'Seoyeon'
    | 'Takumi'
    | 'Tatyana'
    | 'Vicki'
    | 'Vitoria'
    | 'Zeina'
    | 'Zhiyu';
  /**
   * This interface was referenced by `BotConfig`'s JSON-Schema
   * via the `definition` "aws-region-polly".
   */
  export type AwsRegionPolly =
    | 'eu-north-1'
    | 'ap-south-1'
    | 'eu-west-3'
    | 'eu-west-2'
    | 'eu-west-1'
    | 'ap-northeast-2'
    | 'ap-northeast-1'
    | 'sa-east-1'
    | 'ca-central-1'
    | 'ap-southeast-1'
    | 'ap-southeast-2'
    | 'eu-central-1'
    | 'us-east-1'
    | 'us-east-2'
    | 'us-west-1'
    | 'us-west-2';
  /**
   * This interface was referenced by `BotConfig`'s JSON-Schema
   * via the `definition` "bot-channel".
   */
  export type BotChannel =
    | ApiChannel
    | WebchatChannel
    | FacebookChannel
    | KikChannel
    | SlackChannel;
  /**
   * This interface was referenced by `BotConfig`'s JSON-Schema
   * via the `definition` "aws-region".
   */
  export type AwsRegion =
    | 'eu-north-1'
    | 'ap-south-1'
    | 'eu-west-3'
    | 'eu-west-2'
    | 'eu-west-1'
    | 'ap-northeast-2'
    | 'ap-northeast-1'
    | 'sa-east-1'
    | 'ca-central-1'
    | 'ap-southeast-1'
    | 'ap-southeast-2'
    | 'eu-central-1'
    | 'us-east-1'
    | 'us-east-2'
    | 'us-west-1'
    | 'us-west-2';
  /**
   * This interface was referenced by `BotConfig`'s JSON-Schema
   * via the `definition` "dialogflow-api-version".
   */
  export type DialogflowApiVersion = 'v1' | 'v2';
  /**
   * This interface was referenced by `BotConfig`'s JSON-Schema
   * via the `definition` "watson-api-version".
   */
  export type WatsonApiVersion = '2018-02-16' | '2018-09-19';
  /**
   * This interface was referenced by `BotConfig`'s JSON-Schema
   * via the `definition` "bot-engine-settings".
   */
  export type BotEngineSettings =
    | AwsLexEngineSettings
    | DialogflowEngineSettings
    | WatsonEngineSettings
    | OpenaiEngineSettings;
  /**
   * This interface was referenced by `BotConfig`'s JSON-Schema
   * via the `definition` "bot-channel-settings".
   */
  export type BotChannelSettings =
    | ApiChannelSettings
    | WebchatChannelSettings
    | FacebookChannelSettings
    | KikChannelSettings
    | SlackChannelSettings;

  export interface BotConfig {
    name: string;
    team: string;
    engines: BotEngines;
    options: BotOptions;
    channels: BotChannels;
    [k: string]: any;
  }
  /**
   * This interface was referenced by `BotConfig`'s JSON-Schema
   * via the `definition` "bot-engines".
   */
  export interface BotEngines {
    [k: string]: BotEngine;
  }
  /**
   * This interface was referenced by `BotConfig`'s JSON-Schema
   * via the `definition` "aws-lex-engine".
   */
  export interface AwsLexEngine {
    service: 'amazon-lex' | 'aws-lex';
    label?: AdapterLabel;
    hidden?: boolean;
    settings: AwsLexEngineSettings;
    [k: string]: any;
  }
  /**
   * This interface was referenced by `BotConfig`'s JSON-Schema
   * via the `definition` "aws-lex-engine-settings".
   */
  export interface AwsLexEngineSettings {
    botName: string;
    botAlias: string;
    awsRegion: AwsRegionLex;
    awsCredentials: AwsCredentials;
  }
  /**
   * This interface was referenced by `BotConfig`'s JSON-Schema
   * via the `definition` "aws-credentials".
   */
  export interface AwsCredentials {
    awsAccessKeyId: string;
    awsSecretAccessKey: string;
    [k: string]: any;
  }
  /**
   * This interface was referenced by `BotConfig`'s JSON-Schema
   * via the `definition` "dialogflow-engine".
   */
  export interface DialogflowEngine {
    service: 'dialogflow';
    label?: AdapterLabel;
    hidden?: boolean;
    settings: DialogflowEngineSettings;
    [k: string]: any;
  }
  /**
   * This interface was referenced by `BotConfig`'s JSON-Schema
   * via the `definition` "dialogflow-engine-settings-v1".
   */
  export interface DialogflowEngineSettingsV1 {
    apiVersion: 'v1';
    clientToken: string;
  }
  /**
   * This interface was referenced by `BotConfig`'s JSON-Schema
   * via the `definition` "dialogflow-engine-settings-v2".
   */
  export interface DialogflowEngineSettingsV2 {
    apiVersion: 'v2';
    dialogflowCredentials: DialogflowCredentials;
  }
  /**
   * This interface was referenced by `BotConfig`'s JSON-Schema
   * via the `definition` "dialogflow-credentials".
   */
  export interface DialogflowCredentials {
    type?: string;
    project_id: string;
    private_key_id?: string;
    private_key: string;
    client_email: string;
    client_id?: string;
    auth_uri?: string;
    token_uri?: string;
    auth_provider_x509_cert_url?: string;
    client_x509_cert_url?: string;
    [k: string]: any;
  }
  /**
   * This interface was referenced by `BotConfig`'s JSON-Schema
   * via the `definition` "watson-engine".
   */
  export interface WatsonEngine {
    service: 'watson';
    label?: AdapterLabel;
    hidden?: boolean;
    settings: WatsonEngineSettings;
    [k: string]: any;
  }
  /**
   * This interface was referenced by `BotConfig`'s JSON-Schema
   * via the `definition` "watson-engine-settings-v1".
   */
  export interface WatsonEngineSettingsV1 {
    apiVersion: '2018-02-16';
    watsonCredentials: WatsonCredentials;
    workspaceId: string;
  }
  /**
   * This interface was referenced by `BotConfig`'s JSON-Schema
   * via the `definition` "watson-credentials".
   */
  export interface WatsonCredentials {
    apikey: string;
    iam_apikey_description: string;
    iam_apikey_name: string;
    iam_role_crn: string;
    iam_serviceid_crn: string;
    url: string;
    [k: string]: any;
  }
  /**
   * This interface was referenced by `BotConfig`'s JSON-Schema
   * via the `definition` "watson-engine-settings-v2".
   */
  export interface WatsonEngineSettingsV2 {
    apiVersion: '2018-09-19';
    watsonCredentials: WatsonCredentials;
    assistantId: string;
  }
  /**
   * This interface was referenced by `BotConfig`'s JSON-Schema
   * via the `definition` "openai-engine".
   */
  export interface OpenaiEngine {
    service?: 'openai';
    label?: AdapterLabel;
    hidden?: boolean;
    settings?: OpenaiEngineSettings;
    [k: string]: any;
  }
  /**
   * This interface was referenced by `BotConfig`'s JSON-Schema
   * via the `definition` "openai-engine-settings".
   */
  export interface OpenaiEngineSettings {
    model: string;
    systemMessage?: string;
    openaiCredentials: OpenaiCredentials;
  }
  /**
   * This interface was referenced by `BotConfig`'s JSON-Schema
   * via the `definition` "openai-credentials".
   */
  export interface OpenaiCredentials {
    openaiApiKey: string;
    openaiOrganizationId?: string;
    [k: string]: any;
  }
  /**
   * This interface was referenced by `BotConfig`'s JSON-Schema
   * via the `definition` "bot-options".
   */
  export interface BotOptions {
    defaultEngine: string;
    defaultConversationLifetime?: number;
    icon?: string | null;
    supportedSystemEvents?: string[] | null;
    awsVoice?: AwsPollyVoice;
    awsRegion?: AwsRegionPolly;
    awsCredentials?: AwsCredentials;
    [k: string]: any;
  }
  /**
   * This interface was referenced by `BotConfig`'s JSON-Schema
   * via the `definition` "bot-channels".
   */
  export interface BotChannels {
    [k: string]: BotChannel;
  }
  /**
   * This interface was referenced by `BotConfig`'s JSON-Schema
   * via the `definition` "api-channel".
   */
  export interface ApiChannel {
    service: 'api';
    label?: AdapterLabel;
    hidden?: boolean;
    settings: ApiChannelSettings;
  }
  /**
   * This interface was referenced by `BotConfig`'s JSON-Schema
   * via the `definition` "api-channel-settings".
   */
  export interface ApiChannelSettings {
    webhook: string;
    authToken: string;
    shouldSendPreAndPostQueryTypingMessages?: boolean;
    awsVoice?: AwsPollyVoice;
    awsRegion?: AwsRegionPolly;
    awsCredentials?: AwsCredentials;
  }
  /**
   * This interface was referenced by `BotConfig`'s JSON-Schema
   * via the `definition` "webchat-channel".
   */
  export interface WebchatChannel {
    service: 'webchat';
    label?: AdapterLabel;
    hidden?: boolean;
    settings: WebchatChannelSettings;
  }
  /**
   * This interface was referenced by `BotConfig`'s JSON-Schema
   * via the `definition` "webchat-channel-settings".
   */
  export interface WebchatChannelSettings {
    authToken: string;
    awsVoice?: AwsPollyVoice;
    awsRegion?: AwsRegionPolly;
    awsCredentials?: AwsCredentials;
    icon?: string | null;
  }
  /**
   * This interface was referenced by `BotConfig`'s JSON-Schema
   * via the `definition` "facebook-channel".
   */
  export interface FacebookChannel {
    service: 'facebook';
    label?: AdapterLabel;
    hidden?: boolean;
    settings: FacebookChannelSettings;
  }
  /**
   * This interface was referenced by `BotConfig`'s JSON-Schema
   * via the `definition` "facebook-channel-settings".
   */
  export interface FacebookChannelSettings {
    verifyToken: string;
    pageToken: string;
    awsVoice?: AwsPollyVoice;
    awsRegion?: AwsRegionPolly;
    awsCredentials?: AwsCredentials;
  }
  /**
   * This interface was referenced by `BotConfig`'s JSON-Schema
   * via the `definition` "kik-channel".
   */
  export interface KikChannel {
    service: 'kik';
    label?: AdapterLabel;
    hidden?: boolean;
    settings: KikChannelSettings;
  }
  /**
   * This interface was referenced by `BotConfig`'s JSON-Schema
   * via the `definition` "kik-channel-settings".
   */
  export interface KikChannelSettings {
    userName: string;
    apiKey: string;
    manuallySendReadReceipts?: boolean;
    receiveReadReceipts?: boolean;
    receiveDeliveryReceipts?: boolean;
    receiveIsTyping?: boolean;
    awsVoice?: AwsPollyVoice;
    awsRegion?: AwsRegionPolly;
    awsCredentials?: AwsCredentials;
  }
  /**
   * This interface was referenced by `BotConfig`'s JSON-Schema
   * via the `definition` "slack-channel".
   */
  export interface SlackChannel {
    service: 'slack';
    label?: AdapterLabel;
    hidden?: boolean;
    settings: SlackChannelSettings;
  }
  /**
   * This interface was referenced by `BotConfig`'s JSON-Schema
   * via the `definition` "slack-channel-settings".
   */
  export interface SlackChannelSettings {
    botUserToken: string;
    signingSecret: string;
  }
  /**
   * This interface was referenced by `BotConfig`'s JSON-Schema
   * via the `definition` "bots".
   */
  export interface Bots {
    $schema?: string;
    [k: string]: BotConfig;
  }

  export type BotOptionKey = keyof BotOptions;
  export type BotOptionValue<key extends keyof BotOptions> = BotOptions[key];
}
